### DISCUSSION TITLE
Few questions

### DISCUSSION THREAD

**Discussion by qwerty2586** [lainio/err2#4] [Q&A] [answered]
Hello, i am using package for a while in several projects and I am refactoring old code to get rid of underscores. I have stumbled to few issues that are missing from documnetation.

### Simple way to overwrite errors?
Or just annotate single try, so i would be able to recognize type of error in error hadler. And if possible check if annotation is present.

### StacktraceWriter doesnt print annotations. Is this intentional? 
I thought that purpose of StackTraceWriter is to setup and forget about printing errors everywhere.

### Example to use try.Is()
I am not sure how to use it. Also more examples please.

Thank you for library. It helps with writing much better code.

**Comment by lainio** [comment, marked as answer]
Hello, and thank you very much for your comments and questions. They are highly appreciated!

### Simple way to overwrite errors?

The current `err2` package follows the Go2 try-proposal, where you add as many deferred error handlers as your function needs.

If I understand your question right, you would like to know what was the specific `try.To` which triggered the error. The preferred (and only) way is to use as many **deferred** `err2.Handle` or `err2.Returnf`. The order of the handlers tells what to do in the particular error handler. Please see [the following playground](https://go.dev/play/p/bUGX8yDAeoL). There is just one `err2.Handle` for the cleanup. The first one, `err2.Returnf`, annotates the error. Both are called when there is an error after target file creation.

If you wrap the error in any handler, you can use Go's `errors.Is` function to check if the wrapping is already done.

Please see [the following playground example](https://goplay.tools/snippet/KW6PPQnzZCg) about how to 'override' or wrap the current error. The `err2.Handle` handler now replaces the function's first `err2.Returnf` statement and annotates the error slightly differently.

### StackTraceWriter should print annotations. If not, it's not intentional

Where stack tracing is on, please see [this altered playground](https://go.dev/play/p/jIBS6BoGAbn). Please also notice that there are two different stack traces. The highest level of the call stack `main` function naturally has most of the error annotation:
```
---
copy /tmpfs/play_not_exist /file_in_root: open /tmpfs/play_not_exist: no such file or directory
---
```
The annotation is shorter in the `CopyFile` function itself because it hasn't yet annotated the error when the automatic `StackTracing` prints it.
```
---
open /tmpfs/play_not_exist: no such file or directory
---
```
I'm still open to suggestions. For our use cases, this has been enough, and to be honest, it's easiest to implement, and it has been most helpful for debugging.

### Example to use try.Is()

About the examples, please see [the official err2 Go documentation](https://pkg.go.dev/github.com/lainio/err2).

About the `try.Is()`, the following code block is taken directly from our project:
```
	...
	myAE, err := d.AEndp()
	if !try.Is(err, ErrNoData) {
		return NewDoc(d, myAE)
	}
	return NewDoc(d, ae)
```
`try.Is` return true if err is (internally uses Go's `errors.Is`) ErrNoData **and** it also checks the other errors as well, i.e., if any other error happens, it behaves like regular `try.To`.

Summary of `try.Is`: it only return `false` when `err == nil` and if `err == ErrNoData` it returns `true` so we can make a decision.

I hope I could have been a help. If you have any other questions or comments, please let me know.

Best regards,
-Harri
